[Infra] Migrate rest of linux builder workflows off GCP runners.

saienduri opened the pull request:
This commit is part of this larger issue that is tracking our migration off the GCP runners, storage buckets, etc: https://github.com/iree-org/iree/issues/18238.

This builds on https://github.com/iree-org/iree/pull/18381, which migrated
* `linux_x86_64_release_packages`
* `linux_x64_clang_debug`
* `linux_x64_clang_tsan`

Here, we move over the rest of the critical linux builder workflows off of the GCP runners:
* `linux_x64_clang`
* `linux_x64_clang_asan`

This also drops all CI usage of the GCP cache (`http://storage.googleapis.com/iree-sccache/ccache`). Some workflows now use sccache backed by Azure Blob Storage as a replacement. There are few issues with this (https://github.com/mozilla/sccache/issues/2258) that prevent us providing read only access to the cache in PRs created from forks, so **PRs from forks currently don't use the cache and will have slower builds**. We're covering for this slowdown by using larger runners, but if we can roll out caching to all builds then we might use runners with fewer cores.

Along with the changes to the cache, usage of Docker is rebased on images in the https://github.com/iree-org/base-docker-images/ repo and the `build_tools/docker/docker_run.sh` script is now only used by unmigrated workflows (`linux_arm64_clang` and `build_test_all_bazel`).

saienduri committed:
make workflows use cluster

Signed-off-by: saienduri <saimanas.enduri@amd.com>

saienduri committed:
make workflows use cluster

Signed-off-by: saienduri <saimanas.enduri@amd.com>

ScottTodd committed:
Migrate base clang workflow to new Dockerfile. (#18392)

Progress on https://github.com/iree-org/iree/issues/15332. This uses a
new `cpubuilder_ubuntu_jammy_x86_64` dockerfile from
https://github.com/iree-org/base-docker-images.

This stops using the remote cache that is hosted on GCP. Build time
_without a cache_ is about 20 minutes on current runners, while build
_with a cache_ is closer to 10 minutes. Build time without a cache is
closer to 28-30 minutes on new runners. We can try adding back a cache
using GitHub or our own hosted storage.

I tried to continue using the previous cache during this transition
period, but the `gcloud` command needs to run on the host, and I'd like
to stop using the `docker_run.sh` script. I'm hoping we can keep folding
away this sort of complexity by having the build machines run a
dockerfile that includes key environment components like utility tools
and any needed authorization/secrets (see
https://github.com/iree-org/iree/issues/18238).

ci-exactly: linux_x64_clang

ScottTodd committed:
Migrate base clang workflow to new Dockerfile. (#18392)

Progress on https://github.com/iree-org/iree/issues/15332. This uses a
new `cpubuilder_ubuntu_jammy_x86_64` dockerfile from
https://github.com/iree-org/base-docker-images.

This stops using the remote cache that is hosted on GCP. Build time
_without a cache_ is about 20 minutes on current runners, while build
_with a cache_ is closer to 10 minutes. Build time without a cache is
closer to 28-30 minutes on new runners. We can try adding back a cache
using GitHub or our own hosted storage.

I tried to continue using the previous cache during this transition
period, but the `gcloud` command needs to run on the host, and I'd like
to stop using the `docker_run.sh` script. I'm hoping we can keep folding
away this sort of complexity by having the build machines run a
dockerfile that includes key environment components like utility tools
and any needed authorization/secrets (see
https://github.com/iree-org/iree/issues/18238).

ci-exactly: linux_x64_clang
Signed-off-by: saienduri <saimanas.enduri@amd.com>

ScottTodd committed:
Rework how ASan and TSan workflows use Docker. (#18396)

Progress on https://github.com/iree-org/iree/issues/15332. I'm trying to
get rid of the `docker_run.sh` scripts, replacing them with GitHub's
`container:` feature. While local development flows _may_ want to use
Docker like the CI workflows do, those scripts contained a lot of
special handling and file mounting to be compatible with Bazel. Much of
that is not needed for CMake and can be folded away, though the
`--privileged` option needed here is one exception.

This stops using the remote cache that is hosted on GCP. We can try
adding back a cache using GitHub or our own hosted storage as part of
https://github.com/iree-org/iree/issues/18238.

Job | Cache? | Runner cluster | Time | Logs
-- | -- | -- | -- | --
ASan | Cache | GCP runners | 14 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10620030527/job/29438925064)
ASan | No cache | GCP runners | 28 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10605848397/job/29395467181)
ASan | Cache | Azure runners | (not configured yet)
ASan | No cache | Azure runners | 35 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10621238709/job/29442788013?pr=18396)
| | |
TSan | Cache | GCP runners | 12 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10612418711/job/29414025939)
TSan | No cache | GCP runners | 21 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10605848414/job/29395467002)
TSan | Cache | Azure runners | (not configured yet)
TSan | No cache | Azure runners | 32 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10621238738/job/29442788341?pr=18396)

ci-exactly: linux_x64_clang_asan
Signed-off-by: saienduri <saimanas.enduri@amd.com>

ScottTodd committed:
Rework how ASan and TSan workflows use Docker. (#18396)

Progress on https://github.com/iree-org/iree/issues/15332. I'm trying to
get rid of the `docker_run.sh` scripts, replacing them with GitHub's
`container:` feature. While local development flows _may_ want to use
Docker like the CI workflows do, those scripts contained a lot of
special handling and file mounting to be compatible with Bazel. Much of
that is not needed for CMake and can be folded away, though the
`--privileged` option needed here is one exception.

This stops using the remote cache that is hosted on GCP. We can try
adding back a cache using GitHub or our own hosted storage as part of
https://github.com/iree-org/iree/issues/18238.

Job | Cache? | Runner cluster | Time | Logs
-- | -- | -- | -- | --
ASan | Cache | GCP runners | 14 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10620030527/job/29438925064)
ASan | No cache | GCP runners | 28 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10605848397/job/29395467181)
ASan | Cache | Azure runners | (not configured yet)
ASan | No cache | Azure runners | 35 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10621238709/job/29442788013?pr=18396)
| | | 
TSan | Cache | GCP runners | 12 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10612418711/job/29414025939)
TSan | No cache | GCP runners | 21 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10605848414/job/29395467002)
TSan | Cache | Azure runners | (not configured yet)
TSan | No cache | Azure runners | 32 minutes |
[logs](https://github.com/iree-org/iree/actions/runs/10621238738/job/29442788341?pr=18396)

ci-exactly: linux_x64_clang_asan

ScottTodd committed:
Use new cpubuilder dockerfile for ASan and TSan jobs. (#18403)

Following https://github.com/iree-org/base-docker-images/pull/6, the new
cpubuilder dockerfile should have all the software needed for ASan and
TSan building + testing (specifically `clang-19` instead of just
`clang-14`).

Progress on https://github.com/iree-org/iree/issues/15332. The only
remaining uses of `gcr.io/iree-oss/base.*` are:

* `build_test_all_bazel` uses `gcr.io/iree-oss/base-bleeding-edge`
* `publish_website` uses `gcr.io/iree-oss/base`
* arm64 workflows use `gcr.io/iree-oss/base-arm64`
* `gcr.io/iree-oss/emscripten` (used by web test workflows) depends on
`gcr.io/iree-oss/base`

ScottTodd committed:
Use new cpubuilder dockerfile for ASan and TSan jobs. (#18403)

Following https://github.com/iree-org/base-docker-images/pull/6, the new
cpubuilder dockerfile should have all the software needed for ASan and
TSan building + testing (specifically `clang-19` instead of just
`clang-14`).

Progress on https://github.com/iree-org/iree/issues/15332. The only
remaining uses of `gcr.io/iree-oss/base.*` are:

* `build_test_all_bazel` uses `gcr.io/iree-oss/base-bleeding-edge`
* `publish_website` uses `gcr.io/iree-oss/base`
* arm64 workflows use `gcr.io/iree-oss/base-arm64`
* `gcr.io/iree-oss/emscripten` (used by web test workflows) depends on
`gcr.io/iree-oss/base`

Signed-off-by: saienduri <saimanas.enduri@amd.com>

saienduri committed:
change workflows to run on gha runner scale set

Signed-off-by: saienduri <saimanas.enduri@amd.com>

saienduri committed:
change workflows to run on gha runner scale set

Signed-off-by: saienduri <saimanas.enduri@amd.com>

Commit:
added ccache

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Commit:
added ccache

Signed-off-by: Elias Joseph <eljoseph@amd.com>
Signed-off-by: saienduri <saimanas.enduri@amd.com>

Eliasj42 committed:
Implemented caching with Azure containers using sccache (#18466)

Implemented caching with Azure containers using sccache, only works when merging from a branch

ci-exactly: linux_x64_clang

Commit:
added sccache for asan, tsan, debug, and release_packages

Signed-off-by: Elias Joseph <eljoseph@amd.com>

Eliasj42 committed:
Merge branch 'shared/runner-cluster-migration' into shared/runner-cluster-migration-add-sccache-to-workflows

ScottTodd commented on .github/workflows/ci_linux_x64_clang.yml:
```diff
@@ -28,37 +28,49 @@ jobs:
-      - os-family=Linux
+    runs-on: azure-linux-scale
+    container:
+      image: ghcr.io/iree-org/cpubuilder_ubuntu_jammy_x86_64@sha256:2b2ad51d7de988be13086bc618d89d2ba47fbf09eb5b38c60dce82b595fb1c74
```
I'll want to make sure all of these are using the latest image prior to merging. Right now there are a few different versions of the image in use.

(still pinned to a SHA, not the `main` tag)

ScottTodd committed:
Skip sccache when the github secret is not available. (#18499)

This fixes the issue shown here:
https://github.com/iree-org/iree/actions/runs/10800586282/job/29958939877#step:6:16
where `SCCACHE_AZURE_CONNECTION_STRING` is defined to empty string
instead of being undefined. It also introduces a config script, as
discussed here:
https://github.com/iree-org/iree/pull/18489#discussion_r1755277304.

We may still want to limit writing to the cache to `push` events.

ci-exactly: linux_x64_clang

ScottTodd committed:
Update all uses of cpubuilder docker image to the latest SHA, fix lint.

ScottTodd commented on .github/workflows/ci_linux_x64_clang_tsan.yml:
```diff
@@ -32,21 +30,28 @@ jobs:
+          # Use a modern clang explicitly.
+          CC: clang-19
+          CXX: clang++-19
+        run: ./build_tools/cmake/build_and_test_tsan.sh
```
60 second timeout running the `iree/tools/test/compile_to_continuation.mlir.test` test 🤔 

https://github.com/iree-org/iree/actions/runs/10838630188/job/30077305097?pr=18511#step:5:12673

Can see if that keeps happening or if it was a flake. TSan is still a nightly build so lower priority to fix individual tests like that

ScottTodd replied:
At least it succeeded on a retry.

saienduri replied:
Popped up again btw

ScottTodd replied:
Still timing out. Looks like several long tests are quite a bit slower on these runners.

This PR:
https://github.com/iree-org/iree/actions/runs/10852963659/job/30120192675?pr=18511#step:5:12518
```
1455/1469 Test  #987: iree/tests/compiler_driver/streams.mlir.test ......................................................................   Passed   10.64 sec
1456/1469 Test    #5: iree/tools/test/compile_to_phase.mlir.test ........................................................................   Passed   20.14 sec
1457/1469 Test #1227: iree/tests/e2e/stablehlo_models/edge_detection.mlir.test ..........................................................   Passed   11.17 sec
1458/1469 Test #2127: iree/samples/multiple_modules/pipeline_sync.mlir.test .............................................................   Passed    8.78 sec
1459/1469 Test #1226: iree/tests/e2e/stablehlo_models/collatz.mlir.test .................................................................   Passed   11.26 sec
1460/1469 Test  #323: iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_vector_distribute.mlir.test .....................................   Passed   17.47 sec
1461/1469 Test #2126: iree/samples/multiple_modules/pipeline_async.mlir.test ............................................................   Passed    9.55 sec
1462/1469 Test #2115: iree/samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg_two_matmul.mlir.test .......................................   Passed   10.58 sec
1463/1469 Test  #287: iree/compiler/Codegen/LLVMGPU/test/nvvm_pipeline_test.mlir.test ...................................................   Passed   28.12 sec
1464/1469 Test #1230: iree/tests/e2e/stablehlo_models/mnist_fake_weights.mlir.test ......................................................   Passed   22.33 sec
1465/1469 Test #1229: iree/tests/e2e/stablehlo_models/fullyconnected.mlir.test ..........................................................   Passed   25.03 sec
1466/1469 Test #1150: iree/tests/e2e/regression/libm_linking.mlir.test ..................................................................   Passed   25.27 sec
1467/1469 Test #1231: iree/tests/e2e/stablehlo_models/unidirectional_lstm.mlir.test .....................................................   Passed   31.80 sec
1468/1469 Test  #391: iree/compiler/ConstEval/test/jit_globals.mlir.test ................................................................   Passed   44.56 sec
1469/1469 Test    #4: iree/tools/test/compile_to_continuation.mlir.test .................................................................***Timeout  60.01 sec
```

Baseline:
https://github.com/iree-org/iree/actions/runs/10846625133/job/30099971895#step:3:12573
```
1457/1468 Test  #361: iree/compiler/Codegen/SPIRV/test/pipeline_matmul_promotion.mlir.test ..............................................   Passed   10.68 sec
1458/1468 Test #2125: iree/samples/multiple_modules/pipeline_async.mlir.test ............................................................   Passed    3.73 sec
1459/1468 Test #2126: iree/samples/multiple_modules/pipeline_sync.mlir.test .............................................................   Passed    3.77 sec
1460/1468 Test #2114: iree/samples/custom_dispatch/cpu/mlp_plugin/mlp_linalg_two_matmul.mlir.test .......................................   Passed    4.55 sec
1461/1468 Test  #323: iree/compiler/Codegen/LLVMGPU/test/ROCDL/pipeline_vector_distribute.mlir.test .....................................   Passed   12.30 sec
1462/1468 Test #1229: iree/tests/e2e/stablehlo_models/mnist_fake_weights.mlir.test ......................................................   Passed    9.25 sec
1463/1468 Test #1228: iree/tests/e2e/stablehlo_models/fullyconnected.mlir.test ..........................................................   Passed    9.40 sec
1464/1468 Test #1149: iree/tests/e2e/regression/libm_linking.mlir.test ..................................................................   Passed   10.03 sec
1465/1468 Test #1230: iree/tests/e2e/stablehlo_models/unidirectional_lstm.mlir.test .....................................................   Passed   10.73 sec
1466/1468 Test  #287: iree/compiler/Codegen/LLVMGPU/test/nvvm_pipeline_test.mlir.test ...................................................   Passed   16.55 sec
1467/1468 Test  #391: iree/compiler/ConstEval/test/jit_globals.mlir.test ................................................................   Passed   17.91 sec
1468/1468 Test    #4: iree/tools/test/compile_to_continuation.mlir.test .................................................................   Passed   27.82 sec
```

ScottTodd replied:
We could skip those specific tests on TSan or increase the ctest timeout. Looks like we currently set it per test, with no easy way to override or scale across the whole ctest run.

ScottTodd replied:
Increased the timeouts.

saienduri replied:
Thanks Scott!

Eliasj42 committed:
added sccache for asan, tsan, debug (#18489)

| Workflow      | Un-cached      | Cached   |
| ------------- | ------------- | ------------- |
| Clang | 15m57s | 10m12s |
| Clang ASan | NA | 15m7s |
| Clang TSan | NA | 9m42s |
| Clang Debug | 12m19s | 8m15s |

Commit:
Testing runner times from forked PR

ScottTodd commented on test.txt:
Revert before merge. (This should be tested in a separate PR, test commits should not go directly on the branch we intend to merge)

ScottTodd replied:
(Especially if the test commits don't pass the DCO or pre-commit checks)

ScottTodd committed:
Remove stale comments about cache from workflows. (#18522)

Also testing these workflows with PRs from a fork.

saienduri committed:
Revert "Testing runner times from forked PR"

This reverts commit 73886083415cd2b0db51b98390313b70a46a6cc3.

ScottTodd committed:
Increase timeouts for tests that are slow under TSan.

Some of these tests are taking 30-60 seconds on new runner machines under ASan/TSan, getting close to the 60 second timeout. Increase the timeout to 5 minutes.

We could also do something TSan/ASan-specific here, but developers running the tests on slower systems can also benefit from these timeout changes.
